---
output: html_document
params: 
  dataset: ""
  working_folder: ""
  data_file: ""
---

```{r setup_read_data, include=FALSE}

# Make html report for genotyping step
# Alexey Larionov
# 25Aug2016

# Read command line parameters

ds <- params$dataset
wf <- params$working_folder
df <- params$data_file

# Set chunks' options
knitr::opts_chunk$set(fig.width=7, fig.height=4, fig.align = 'center', fig.path=wf, echo=FALSE)

# Read data
x <- read.table(df, header=TRUE)
attach(x)
passed.min.filters <- FILTER == "PASS"

# Count numbers of variants
n.all <- format(dim(x)[1], big.mark = ",", trim=TRUE, scientific = FALSE)
n.passed <- format(sum(passed.min.filters), big.mark = ",", trim=TRUE, scientific = FALSE)

```

<center> <H1> `r ds` </H1> </center> 
<center> <H3> Total number of variants emitted by HC: `r n.all` </H3> </center> 
<center> <H4> Number of variants passed minimal filters: `r n.passed` </H4> </center> 
<center> <H4> Minimal HC filter: QUAL > calling threshold (default 30, not emitted) </H4> </center> 
<center> <H4> Minimal VQSR filter: VQSLOD > 0 </H4> </center> 
<br>

```{r counts_tables}

library(printr, lib='/scratch/medgen/tools/r/R-3.2.2/lib64/R/library/')

# Filters summary
n.types.all <- format(sum(table(FILTER, TYPE)), big.mark = ",", trim=TRUE, scientific = FALSE)
c.types.all <- paste("Filters summary (all variants in the file, n=",n.types.all,")", sep ="")
knitr::kable(table(FILTER, TYPE), format.args=list(big.mark=","), caption=c.types.all)

```

---

```{r SNP_histograms_no_filters}
hist(VQSLOD[TYPE == "SNP"], main="SNPs VQSLOD (all, no zoom)", xlab="VQSLOD")

hist(VQSLOD[VQSLOD >=-30 & VQSLOD <= 30 & TYPE == "SNP"], breaks=100, 
     main="SNPs VQSLOD (all, zoom -30 to 30)", xlim=c(-30,30), xlab="VQSLOD")

```

---

```{r SNP_histograms_after_min_filters}
hist(VQSLOD[TYPE == "SNP" & passed.min.filters], main="SNPs VQSLOD (passed min filters, no zoom)", xlab="VQSLOD")

hist(VQSLOD[VQSLOD >=-30 & VQSLOD <= 30 & TYPE == "SNP" & passed.min.filters], breaks=100, 
     main="SNPs VQSLOD (passed min filters, zoom -30 to 30)", xlim=c(-30,30), xlab="VQSLOD")

```

---

```{r INDEL_histograms_no_filters}
hist(VQSLOD[TYPE == "INDEL"], main="INDELs VQSLOD (all, no zoom)", xlab="VQSLOD")

hist(VQSLOD[VQSLOD >=-30 & VQSLOD <= 30 & TYPE == "INDEL"], breaks=100, 
     main="INDELs VQSLOD (all, zoom -30 to 30)", xlim=c(-30,30), xlab="VQSLOD")

```

---

```{r INDEL_histograms_after_min_filters}
hist(VQSLOD[TYPE == "INDEL" & passed.min.filters], main="INDELs VQSLOD (passed min filters, no zoom)", xlab="VQSLOD")

hist(VQSLOD[VQSLOD >=-30 & VQSLOD <= 30 & TYPE == "INDEL" & passed.min.filters], breaks=25, 
     main="INDELs VQSLOD (passed min filters, zoom -30 to 30)", xlim=c(-30,30), xlab="VQSLOD")

```

---

```{r DP_histograms_no_filters}
hist(DP, main="DP (all, no zoom)", xlab="DP")

DP.cap <- floor(quantile(DP, 0.99)) # exclude 1% of excessive dp
DP.zm.main <- paste("DP (all, zoom 0 - ",DP.cap,")", sep="")
hist(DP[DP < DP.cap], breaks=100, xlim=c(0,DP.cap), main=DP.zm.main, xlab="DP")
```

---

```{r DP_histograms_after_min_filters}
hist(DP[passed.min.filters], main="DP (passed min filters, no zoom)", xlab="DP")

DP.cap <- floor(quantile(DP, 0.99)) # exclude 1% of excessive dp
DP.zm.main <- paste("DP (passed min filters, zoom 0 - ",DP.cap,")", sep="")
hist(DP[DP < DP.cap & passed.min.filters], breaks=100, xlim=c(0,DP.cap), main=DP.zm.main, xlab="DP")
```

---

```{r QUAL_histograms_no_filters}
hist(QUAL, main="QUAL (all, no zoom)", xlab="QUAL")
hist(QUAL[QUAL < 250], breaks=100, xlim=c(0,250), main="QUAL (all, zoom 0 - 250)", xlab="QUAL")
```

---

```{r QUAL_histograms_after_min_filters}
hist(QUAL, main="QUAL (passed min filters, no zoom)", xlab="QUAL")
hist(QUAL[QUAL < 250 & passed.min.filters], breaks=100, xlim=c(0,250), main="QUAL (passed min filters, zoom 0 - 250)", xlab="QUAL")
```

```{r Clenaup}
detach(x)
```
